src/protocol/ascii.rs

```diff
@@ -148,58 +162,95 @@ impl AsciiProtocol<Stream> {
-                exptime = options.exptime,
-                vlen = value.get_length(),
-                cas = options.cas.unwrap(),
-                noreply = noreply
-            )?;
-        } else {
-            write!(
-                self.reader.get_mut(),
-                "{command} {key} {flags} {exptime} {vlen}{noreply}\r\n",
-                command = command,
-                key = key,
-                flags = value.get_flags(),
-                exptime = options.exptime,
-                vlen = value.get_length(),
-                noreply = noreply
-            )?;
-        }
+        let mut sent_count = 0;
+
+        {
+            let reader = self.reader.get_mut();
+            for (key_ref, value) in entries.into_iter() {
+                let key = key_ref.as_ref();
+                check_key_len(key)?;
+                if options.cas.is_some() {
+                    write!(
+                        reader,
+                        "{command} {key} {flags} {exptime} {vlen} {cas}{noreply}\r\n",
+                        command = command,
+                        key = key,
+                        flags = value.get_flags(),
+                        exptime = options.exptime,
+                        vlen = value.get_length(),
+                        cas = options.cas.unwrap(),
+                        noreply = noreply
+                    )?;
+                } else {
+                    write!(
+                        reader,
+                        "{command} {key} {flags} {exptime} {vlen}{noreply}\r\n",
+                        command = command,
+                        key = key,
+                        flags = value.get_flags(),
+                        exptime = options.exptime,
+                        vlen = value.get_length(),
+                        noreply = noreply
+                    )?;
+                }
 
-        value.write_to(self.reader.get_mut())?;
-        self.reader.get_mut().write(b"\r\n")?;
-        self.reader.get_mut().flush()?;
+                value.write_to(reader)?;
+                reader.write(b"\r\n")?;
+                sent_count += 1;
+            }
+
+            // Flush now that all the requests have been written.
+            reader.flush()?;
+        }
 
         if options.noreply {
             return Ok(true);
         }
 
-        self.reader.read_line(|response| {
-            let response = MemcacheError::try_from(response)?;
-            match response {
-                "STORED\r\n" => Ok(true),
-                "NOT_STORED\r\n" => Ok(false),
-                "EXISTS\r\n" => Err(CommandError::KeyExists)?,
-                "NOT_FOUND\r\n" => Err(CommandError::KeyNotFound)?,
-                response => Err(ServerError::BadResponse(Cow::Owned(response.into())))?,
+        // Receive all the responses. If there were errors, return the first.
+
+        let mut all_stored = true;
+        let mut error_list: Vec<MemcacheError> = Vec::new();
+        for _ in 0..sent_count {
```

Comment:
I haven't tried, but collecting into a `Result` might be more readable here. [example](https://stackoverflow.com/a/26370894/4129373)

Reply:
I'll simplify. Thanks for the pointer.